api/src/main/java/com/graphhopper/util/shapes/Polygon.java

```diff
@@ -17,117 +17,72 @@
-                }
-            }
-        }
 
-        minLat -= growFactor;
-        minLon -= growFactor;
-        maxLat += growFactor;
-        maxLon += growFactor;
-
-        epsilon = (maxLat - minLat) / 10;
-    }
-
-    public Polygon(double[] lat, double[] lon) {
-        this(lat, lon, 0);
-    }
-
-    /**
-     * Lossy conversion to a GraphHopper Polygon.
-     */
-    public static Polygon create(org.locationtech.jts.geom.Polygon polygon) {
-        double[] lats = new double[polygon.getNumPoints()];
-        double[] lons = new double[polygon.getNumPoints()];
-        for (int i = 0; i < polygon.getNumPoints(); i++) {
-            lats[i] = polygon.getCoordinates()[i].y;
-            lons[i] = polygon.getCoordinates()[i].x;
+        Coordinate[] coordinates = new Coordinate[lats.length + 1];
+        for (int i = 0; i < lats.length; i++) {
+            coordinates[i] = new Coordinate(lons[i], lats[i]);
         }
-        return new Polygon(lats, lons);
+        coordinates[lats.length] = coordinates[0];
+        this.polygon = new PreparedPolygon(factory.createPolygon(new PackedCoordinateSequence.Double(coordinates, 2)));
+        this.envelope = polygon.getGeometry().getEnvelopeInternal();
     }
 
-    /**
-     * Wrapper method for {@link Polygon#contains(double, double)}.
-     */
-    public boolean contains(GHPoint point) {
-        return contains(point.lat, point.lon);
+    public static Polygon create(org.locationtech.jts.geom.Polygon polygon) {
+        return new Polygon(new PreparedPolygon(polygon));
     }
 
     @Override
     public boolean intersects(Shape o) {
-        throw new UnsupportedOperationException("Not supported yet.");
+        throw new IllegalArgumentException("Polygon.intersects: shape not supported");
+    }
+
+    public boolean intersects(PointList pointList) {
+        return polygon.intersects(pointList.getCachedLineString(factory, false));
     }
 
     /**
-     * Implements the ray casting algorithm
-     * Code is inspired from here: http://stackoverflow.com/a/218081/1548788
+     * Does the point in polygon check.
      *
      * @param lat Latitude of the point to be checked
      * @param lon Longitude of the point to be checked
      * @return true if point is inside polygon
      */
     public boolean contains(double lat, double lon) {
-        if (lat < minLat || lat > maxLat || lon < minLon || lon > maxLon) {
-            return false;
-        }
-
-        double rayStartLat = maxLat - (minLat / 2);
-        double rayStartLon = minLon - epsilon;
-
-        boolean inside = false;
-        int len = this.lat.length;
-        for (int i = 0; i < len; i++) {
-            if (edgesAreIntersecting(rayStartLon, rayStartLat, lon, lat, this.lon[i], this.lat[i], this.lon[(i + 1) % len], this.lat[(i + 1) % len]))
-                inside = !inside;
-        }
-        return inside;
-
+        return polygon.contains(factory.createPoint(new Coordinate(lon, lat)));
```

Comment:
We could check the envelope first.

Reply:
This should be done from JTS.

Reply:
This should be done from JTS.

Reply:
Ah, i missed that `polygon` is already a `PreparedGeometry`. I would recommend to rename it accordingly e.g `prepPolygon`